kakao/src/main/kotlin/com/agoda/kakao/intercept/Interceptor.kt

```diff
@@ -0,0 +1,141 @@
+         * @param interceptor lambda with intercepting logic
+         */
+        fun onPerform(isOverride: Boolean = false, interceptor: (T, P) -> Unit) {
+            onPerform = isOverride to interceptor
+        }
+
+        /**
+         * Sets the interceptor for the `check` and `perform` operations for a given interaction.
+         * If overridden, breaks the call chain of operation and transfers the responsibility
+         * to invoke the Espresso on the developer.
+         *
+         * This interceptor is prioritized and is being invoked first for both operations.
+         *
+         * @param isOverride if `true` - breaks the call chain, false otherwise
+         * @param interceptor lambda with intercepting logic
+         */
+        fun onAll(isOverride: Boolean = false, interceptor: (T) -> Unit) {
+            onAll = isOverride to interceptor
+        }
+
+        internal fun build(): Interceptor<T, C, P> = Interceptor(onCheck, onPerform, onAll)
+    }
+
+    /**
+     * Configuration class that is used for building interceptors on the
+     * [Kakao][com.agoda.kakao.Kakao] runtime and [Screen][com.agoda.kakao.screen.Screen] levels.
+     *
+     * @see com.agoda.kakao.Kakao
+     * @see com.agoda.kakao.screen.Screen
+     */
+    class Configurator {
+        private var vi: Interceptor<ViewInteraction, ViewAssertion, ViewAction>? = null
+        private var di: Interceptor<DataInteraction, ViewAssertion, ViewAction>? = null
+        private var wi: Interceptor<Web.WebInteraction<*>, WebAssertion<*>, Atom<*>>? = null
+
+        /**
+         * Setups the interceptor for `check` and `perform` operations happening through [ViewInteraction]
+         *
+         * @param builder Builder of interceptor for [ViewInteraction]
+         */
+        fun onViewInteraction(builder: Builder<ViewInteraction, ViewAssertion, ViewAction>.() -> Unit) {
+            vi = Builder<ViewInteraction, ViewAssertion, ViewAction>().apply(builder).build()
+        }
+
+        /**
+         * Setups the interceptor for `check` and `perform` operations happening through [DataInteraction]
+         *
+         * @param builder Builder of interceptor for [DataInteraction]
+         */
+        fun onDataInteraction(builder: Builder<DataInteraction, ViewAssertion, ViewAction>.() -> Unit) {
+            di = Builder<DataInteraction, ViewAssertion, ViewAction>().apply(builder).build()
+        }
+
+        /**
+         * Setups the interceptor for `check` and `perform` operations happening through [Web.WebInteraction]
+         *
+         * @param builder Builder of interceptor for [Web.WebInteraction]
+         */
+        fun onWebInteraction(builder: Builder<Web.WebInteraction<*>, WebAssertion<*>, Atom<*>>.() -> Unit) {
+            wi = Builder<Web.WebInteraction<*>, WebAssertion<*>, Atom<*>>().apply(builder).build()
+        }
+
+        internal fun configure(): Triple<
```

Comment:
Maybe a data class can be used for this? For readability sake

Reply:
agree

Reply:
Will be pushed soon